server/sqlstore/incident.go

```
@@ -606,23 +606,28 @@ func (s *incidentStore) ChangeCreationDate(incidentID string, creationTimestamp
 
 // HasViewed returns true if userID has viewed channelID
 func (s *incidentStore) HasViewedChannel(userID, channelID string) bool {
-	query := s.queryBuilder.
-		Select("COUNT(*)").
-		From("IR_ViewedChannel vc").
-		Where(sq.And{sq.Eq{"vc.ChannelID": channelID}, sq.Eq{"vc.UserID": userID}})
-
-	var count int
-	err := s.store.getBuilder(s.store.db, &count, query)
+	query := sq.Expr(
+		`SELECT EXISTS(SELECT *
+                         FROM IR_ViewedChannel as vc
+                        WHERE vc.ChannelID = ?
+                          AND vc.UserID = ?)
+             `, channelID, userID)
+
+	var exists bool
+	err := s.store.getBuilder(s.store.db, &exists, query)
 	if err != nil {
 		return false
 	}
 
-	return count > 0
+	return exists
 }
 
-// SetViewed records that userID has viewed channelID. NOTE: does not check if there is already a
-// record of that userID/channelID (i.e., will create duplicate rows)
+// SetViewed records that userID has viewed channelID.
 func (s *incidentStore) SetViewedChannel(userID, channelID string) error {
+	if s.HasViewedChannel(userID, channelID) {
```

nit: This will catch many of the duplicate inserts, but not all of them. Another option is to `Insert` with a condition using the same query as `HasViewedChannel`. But even that won't guarantee anything, and one usually has to explicitly handle the unique constraint exception. 0/5 here, just commenting.

---

@lieut-data Could you clarify why it wouldn't catch duplicate inserts? (Other than the obvious race condition possibility.) But regardless, we don't even technically need that check (or a conditional statement) -- a duplicate entry should return an error. (Correct me if I'm wrong, but that's what I've seen from the tests.)
I've made a commit so that returning an error (for a duplicate entry) is handled correctly.